File: src/explore-education-statistics-admin/src/services/boundaryDataService.ts
import client from '@admin/services/utils/service';
import { LocationLevelKey } from '@common/utils/locationLevelsMap';

export interface BoundaryLevel {
  id: string;
  level: LocationLevelKey;
  label: string;
  published: Date;
}

const boundaryDataService = {
  getBoundaryLevels(): Promise<BoundaryLevel[]> {
    return client.get('/boundary-level');
  },

  getBoundaryLevel(id: string): Promise<BoundaryLevel> {
    return client.get(`/boundary-level/${id}`);
  },

  updateBoundaryLevel(data: {
    id: string;
    label: string;
  }): Promise<BoundaryLevel> {
    return client.patch('/boundary-level', data);
  },

  uploadBoundaryFile(data: {
    level: string;
    label: string;
    published: Date;
    file: File;
  }): Promise<BoundaryLevel[]> {
    const formData = new FormData();
    formData.append('level', data.level);
    formData.append('label', data.label);
    formData.append('file', data.file);
    formData.append('published', data.published.toISOString());

    return client.post('/boundary-level', formData);
  },
};

export default boundaryDataService;

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/BoundaryDataService.cs
#nullable enable
using GeoJSON.Net.Feature;
using GovUk.Education.ExploreEducationStatistics.Admin.Services.Interfaces;
using GovUk.Education.ExploreEducationStatistics.Common.Model.Data;
using GovUk.Education.ExploreEducationStatistics.Common.Utils;
using GovUk.Education.ExploreEducationStatistics.Data.Model;
using GovUk.Education.ExploreEducationStatistics.Data.Model.Repository.Interfaces;
using System;
using System.Linq;
using System.Threading.Tasks;

namespace GovUk.Education.ExploreEducationStatistics.Admin.Services;

public class BoundaryDataService(
    IBoundaryLevelRepository boundaryLevelRepository,
    IBoundaryDataRepository boundaryDataRepository) : IBoundaryDataService
{
    public async Task ProcessGeoJson(
        GeographicLevel level,
        string label,
        DateTime published,
        FeatureCollection featureCollection)
    {
        var boundaryLevel = await boundaryLevelRepository.CreateBoundaryLevel(level, label, published);
        var boundaryData = featureCollection.Features
            .Select(feature =>
            {
                return new BoundaryData
                {
                    BoundaryLevel = boundaryLevel,
                    Code = BoundaryDataUtils.GetCode(feature.Properties),
                    Name = BoundaryDataUtils.GetName(feature.Properties),
                    GeoJson = feature
                };
            }).ToList();

        await boundaryDataRepository.AddRange(boundaryData);
    }
}
